### PR TITLE
Add Quantization recipe filter per operator type for x86_inductor_quantizer

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122777
* #122776
* __->__ #122775

**Summary**
Default recipes are enabled in `X86InductorQuantizer` and request comes to customize recipes based on these defaults.

- Avoid annotation propagation and restrict annotation only to annotate `conv`/`linear`.
- Add `matmul`  in the quantization recipes, noting that it's not a general recipe but tailored to meet accuracy criteria for specific models.

To meet these requests, we made changes in this PR by introducing interface as `set_function_type_qconfig` and `set_module_type_qconfig`

- `set_function_type_qconfig` accepts functional input as `torch.nn.functional.linear` or `torch.matmul`; `set_module_type_qconfig` accepts nn.Module input as `torch.nn.Conv2d`.
- To disable the recipe for this operator, user can simply exclude it from the list of operations as `quantizer.set_function_type_qconfig(op, None)`.
- To modify or extend the recipe for this operator with default recipe, user can customize as `quantizer.set_function_type_qconfig(op, config)`.

**Test Plan**
```
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_filter_conv2d_recipe
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_filter_linear_recipe
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_filter_maxpool2d_recipe
```

### CODE REVIEW

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 1]
```diff
@@ -270,9 +284,32 @@ def set_global(self, quantization_config: QuantizationConfig):
     def set_config_for_operator_type(
         self, operator_type: str, quantization_config: QuantizationConfig
     ):
-        self.operator_type_config[operator_type] = quantization_config
+        warnings.warn(
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:287]: you can just remove if you don't have users, the original intention for this is to allow people to set things like: "linear", "linear_relu", ...

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 2]
```diff
@@ -270,9 +284,32 @@ def set_global(self, quantization_config: QuantizationConfig):
...
+            "The function set_config_for_operator_type has been deprecated. Please use set_operator_type instead."
+        )
+
+    def set_operator_type(
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:291]: I think it's probably better to stay in the user facing things like `nn.Linear`, 'nn.functional.linear` instead of aten ops, since modeling user are not interacting with aten ops

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py:287]: Thanks and removed.

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py:291]: Thanks for the comment and agree. How about we:

- Change `set_operator_type` as a internal interface as `_set_aten_operator_type` and `_get_aten_operator_type` for implementation.
- Add new user-face API like `set_module_type` and `set_function_type` to accept parameters of `nn.Linear` as module and `nn.funcational.linear` or `torch.matmul` as function.

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 3]
```diff
@@ -403,36 +433,30 @@ def _annotate_for_static_quantization_config(
...
+        self._annotate_conv2d_fusion_pattern(model)
+        self._annotate_linear_fusion_pattern(model)
 
         # Step2: Recipe to propagate annotation for patterns beside conv/linear.
         # Go through all the nodes from start to end.
         # Recipe refer to https://github.com/intel/intel-extension-for-pytorch/blob/
         # 90d19323d96afc53fcc22ba5a7bb3fb07fdd6c1c/intel_extension_for_pytorch/quantization/_recipe.py#L538
         for node in model.graph.nodes:
-            self._annotation_propagation_quantizable_pattern(node, config)
+            self._annotation_propagation_quantizable_pattern(node)
```
Comment (jgong5) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:445]: nit
```suggestion
            self._annotate_propagation_quantizable_pattern(node)
```

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 4]
```diff
@@ -267,12 +281,28 @@ def set_global(self, quantization_config: QuantizationConfig):
...
+    ) -> "X86InductorQuantizer":
+        if operator_type in quantizable_ops:
+            self.operator_type_config[operator_type] = quantization_config
+        else:
+            warnings.warn(
+                f"operator: {operator} is not supported to do quantization by X86InductorQuantizer."
+            )
         return self
 
+    def _get_aten_operator_type(
```
Comment (jgong5) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:297]:
```suggestion
    def _get_aten_operator_qconfig(
```

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 5]
```diff
@@ -267,12 +281,28 @@ def set_global(self, quantization_config: QuantizationConfig):
         self.global_config = quantization_config
         return self
 
-    def set_config_for_operator_type(
-        self, operator_type: str, quantization_config: QuantizationConfig
-    ):
-        self.operator_type_config[operator_type] = quantization_config
+    def _set_aten_operator_type(
```
Comment (jgong5) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:284]:
```suggestion
    def _set_aten_operator_qconfig(
```

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 6]
```diff
@@ -267,12 +281,28 @@ def set_global(self, quantization_config: QuantizationConfig):
...
+        self,
+        operator_type: torch._ops.OpOverloadPacket,
+        quantization_config: Optional[QuantizationConfig],
+    ) -> "X86InductorQuantizer":
+        if operator_type in quantizable_ops:
+            self.operator_type_config[operator_type] = quantization_config
+        else:
+            warnings.warn(
+                f"operator: {operator} is not supported to do quantization by X86InductorQuantizer."
```
Comment (jgong5) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:293]:
```suggestion
                f"operator: Unable to quantize {operator} by X86InductorQuantizer."
```

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 7]
```diff
@@ -267,12 +281,28 @@ def set_global(self, quantization_config: QuantizationConfig):
...
-        self, operator_type: str, quantization_config: QuantizationConfig
-    ):
-        self.operator_type_config[operator_type] = quantization_config
+    def _set_aten_operator_type(
+        self,
+        operator_type: torch._ops.OpOverloadPacket,
+        quantization_config: Optional[QuantizationConfig],
+    ) -> "X86InductorQuantizer":
+        if operator_type in quantizable_ops:
+            self.operator_type_config[operator_type] = quantization_config
```
Comment (jgong5) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:290]:
```suggestion
            self.operator_type_qconfig[operator_type] = quantization_config
```

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 4 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py:297]: Thanks and changed.

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 5 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py:284]: Thanks and changed

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 3 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py:445]: Thanks and changed.

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 7 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py:290]: Thanks and changed.

**Review by jgong5 — APPROVED** [review]

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py]: this sounds good, please update summary as well.

please add the user-facing API as well, although this is starting to look like the old qconfig mapping API, but you get to choose what API to expose for users this time

Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py]: Thanks for the suggestion @jerryzh168. Please kindly help to take a look again.

- Introducing user-facing API as `set_function_type_qconfig` and `set_module_type_qconfig`.
- Summary has been updated.

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py [hunk 8]
```diff
@@ -267,12 +310,62 @@ def set_global(self, quantization_config: QuantizationConfig):
...
+    ) -> "X86InductorQuantizer":
+        if function_type in X86InductorQuantizer.module_function_to_aten_operator_type:
+            self._set_aten_operator_qconfig(
+                X86InductorQuantizer.module_function_to_aten_operator_type[
+                    function_type
+                ],
+                quantization_config,
+            )
+        else:
+            warnings.warn(
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/x86_inductor_quantizer.py:326]: should this just be an error?

**Review by jerryzh168 — APPROVED** [review]
LG, had one comment inline

File: torch/ao/quantization/quantizer/x86_inductor_quantizer.py (hunk 8 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/x86_inductor_quantizer.py]: Thanks and changed.
